burn-import/src/onnx/state.rs

```diff
@@ -0,0 +1,246 @@
+            _record_settings: PhantomData::default(),
+        }
+    }
+}
+
+impl<RS: RecordSettings> Record for ModelState<RS> {
+    type Item<S: RecordSettings> = ModelState<RS>;
+
+    fn into_item<S: RecordSettings>(self) -> Self::Item<S> {
+        self
+    }
+
+    fn from_item<S: RecordSettings>(_item: Self::Item<S>) -> Self {
+        unimplemented!()
+    }
+}
+
+impl<RS: RecordSettings> Serialize for ModelState<RS> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // Collect all the stateful nodes
+        let stateful_nodes: Vec<_> = self
+            .graph
+            .nodes
+            .iter()
+            .filter(|node| node.is_stateful)
+            .collect();
+
+        // Create a map to serialize the stateful nodes
+        let mut map = serializer.serialize_map(Some(stateful_nodes.len()))?;
+
+        // Serialize each stateful node into a map entry
+        for node in stateful_nodes.into_iter() {
+            match node.node_type {
+                NodeType::Conv2d => {
+                    let (name, record_item) = conv2d_state::<RS>(node);
+                    map.serialize_entry(&name, &record_item)?;
+                }
+                NodeType::Linear => {
+                    let (name, record_item) = linear_state::<RS>(node);
+                    map.serialize_entry(&name, &record_item)?;
+                }
+                NodeType::BatchNormalization => match first_input_dim(node).unwrap() {
+                    2 => {
+                        let (name, record_item) = batch_norm_state::<0, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    3 => {
+                        let (name, record_item) = batch_norm_state::<1, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    4 => {
+                        let (name, record_item) = batch_norm_state::<2, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    5 => {
+                        let (name, record_item) = batch_norm_state::<3, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    dim => todo!("BatchNorm for dim = {dim} is not implemented yet"),
+                },
+
+                _ => todo!(
+                    "Serialize state for node type {:?} is not implemented yet",
+                    node.node_type
+                ),
+            }
+        }
+        map.end()
+    }
+}
+
+/// Dummy implementation of Deserialize for ModelState
+impl<'de, RS: RecordSettings> Deserialize<'de> for ModelState<RS> {
+    fn deserialize<D>(_deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        unimplemented!() // No intention to deserialize a ModelState
+    }
+}
```

Comment:
I am forced to implement Deserialize because record for serialization requires it. We are not de-serializing it back to ModelState, but rather we de-serializing to Model struct, which implements all proper methods. So basically I am side stepping the usual steps.

Reply:
Would it be possible to implement deserialize for ModelState and instead implement a From to the model struct? It's just confusing to serialize a struct and deserialize it for another struct.

Reply:
I am not sure how to make it look better. I struggled to get working from the beginning if you recall. We should definitely make it easy to load data from a file not directly generated from the serialization itself. That's why you're seeing consequences. I think it's okay to leave this for now since it's under a rag and come back with a better facility to load the data. This loading exercise was good since it revealed difficulties with the current Burn API.

Reply:
What I mean is the model state can be loaded from disk, and you can load the model struct's weights directly from the model state, no deserialization needed. We can open an issue to resolve this later, but we must cleanup how the model state (de)serialization works.